cron.go

```diff
@@ -170,129 +192,220 @@ func New(opts ...CronOpt) (*Cron, error) {
-	entry := &Entry{
-		Schedule: schedule,
-		Job:      job,
+func (c *Cron) scheduleJob(job Job) error {
+	s, err := Parse(job.Rhythm)
+	if err != nil {
+		return err
 	}
-	if !c.running {
-		c.entriesMutex.Lock()
-		c.entries = append(c.entries, entry)
-		c.entriesMutex.Unlock()
-		return
+
+	return c.schedule(s, job)
+}
+
+// Schedule adds a Job to the Cron to be run on the given schedule.
+func (c *Cron) schedule(schedule Schedule, job Job) error {
+	partitionId := c.partitioning.CalculatePartitionId(job.Name)
+	if !c.partitioning.CheckPartitionLeader(partitionId) {
+		// It means the partitioning changed and persisted jobs are in the wrong partition now.
+		return fmt.Errorf("host does not own partition %d", partitionId)
+	}
+
+	entry := &Entry{
+		Schedule:        schedule,
+		Job:             job,
+		Prev:            time.Unix(0, 0),
+		distMutexPrefix: c.organizer.TicksPath(partitionId) + "/",
 	}
 
-	c.add <- entry
+	c.appendOperation(func(ctx context.Context) *Entry {
+		c.entries[entry.Job.Name] = entry
+		return entry
+	})
+	return nil
+}
+
+func (c *Cron) appendOperation(op func(ctx context.Context) *Entry) {
+	c.operationsMutex.Lock()
+	defer c.operationsMutex.Unlock()
+
+	c.pendingOperations = append(c.pendingOperations, op)
 }
 
 // Entries returns a snapshot of the cron entries.
 func (c *Cron) Entries() []*Entry {
 	if c.running {
-		c.entriesMutex.RLock()
-		defer c.entriesMutex.RUnlock()
-
 		c.snapshot <- nil
 		x := <-c.snapshot
 		return x
 	}
-	return c.entrySnapshot()
+
+	c.operationsMutex.RLock()
+	defer c.operationsMutex.RUnlock()
+	entries := []*Entry{}
+	for _, e := range c.entries {
+		entries = append(entries, &Entry{
+			Schedule: e.Schedule,
+			Next:     e.Next,
+			Prev:     e.Prev,
+			Job:      e.Job,
+		})
+	}
+	return entries
 }
 
 // Start the cron scheduler in its own go-routine.
 func (c *Cron) Start(ctx context.Context) {
 	c.running = true
-	go c.run(ctx)
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	c.cancel = cancel
+	go c.run(ctxWithCancel)
 }
 
 // Run the scheduler.. this is private just due to the need to synchronize
 // access to the 'running' state variable.
 func (c *Cron) run(ctx context.Context) {
+	c.runWaitingGroup.Add(1)
+	mutexStore := NewMutexStore(c.etcdclient)
 	// Figure out the next activation times for each entry.
 	now := time.Now().Local()
-	for _, entry := range c.entries {
-		entry.Next = entry.Schedule.Next(now)
-	}
+
+	changed := make(chan bool)
+	entries := []*Entry{}
+
+	go func(ctx context.Context) {
+		for {
+			hasPendingOperations := false
+			c.operationsMutex.RLock()
+			hasPendingOperations = len(c.pendingOperations) > 0
+			c.operationsMutex.RUnlock()
+
+			if hasPendingOperations {
+				changed <- true
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(time.Second):
+			}
+		}
+	}(ctx)
 
 	for {
-		// Determine the next entry to run.
-		sort.Sort(byTime(c.entries))
+		sort.Sort(byTime(entries))
```

Comment:
It is extremely expensive to be looping over every entry every second to be checking the time, and re-ordering. Is that what we doing?

It seems like a time based queue fits better here- we gain both non-busy waiting, more precision and less ranging over data. It also isn't clear what currently happens if triggering all events takes longer than a tick. We have a queue implemented in kit and is being used for actor timers for the same purpose.

https://github.com/dapr/kit/blob/main/events/queue/processor.go
https://github.com/dapr/dapr/blob/7c3011e7817dcc5b68ff9769964c2982b4102354/pkg/actors/timers/timers.go#L108